Toolkit analysis (#60)
mc2-center/cckp-toolkit-workflow: analysis/README.md

# Analysis code: "Mapping the Garden"

> Reproducible analysis for the manuscript *Mapping the Garden: Software Sustainability as a

Reviewer: Consider following one sentence per line structure with markdown documents to help make versioning and review more streamlined. When markdown is rendered each line without an additional break is rendered as a single paragraph. There's a tool here which might help too: [onesentence](https://github.com/CU-DBMI/onesentence) .

> Correlate of Scientific Impact Across Computational Biology Ecosystems*. These scripts take the

Reviewer: Generally I'd guide towards avoiding the use of "script" unless referencing an individual, standalone Python file.
```suggestion
Correlate of Scientific Impact Across Computational Biology Ecosystems*. This analysis uses the
```

per-repository measurements produced by the Cancer Complexity Toolkit (CCT) Nextflow pipeline (in
the repository root) and produce the statistics and figures in the paper.

## Setup

```bash
# from the repository root
python3 -m venv .venv && source .venv/bin/activate

> pip install -r analysis/requirements.txt

Reviewer: Consider using a `pyproject.toml` file and leveraging [`uv`](https://docs.astral.sh/uv/) to help both run and document the dependencies associated with the analysis (see the [migration guide here](https://pydevtools.com/handbook/how-to/migrate-requirements.txt/)). This becomes especially helpful with lockfiles like the `uv.lock` file, which automatically can freeze the precise dependency versions last used for the project.

```

> All scripts are run **from the repository root** and use argparse with sensible defaults. Pass

Reviewer: ```suggestion
All code is processed **from the repository root** and use argparse with sensible defaults. Pass
```

`--help` to any script to see its options.

## Data

> The scripts read from and write to `data/final_results/`. That directory is **not committed**

Reviewer: ```suggestion
The code reads from and writes to `data/final_results/`. That directory is intentionally **ignored**
```

> (it holds large derived tables); it is released separately with the manuscript (see the Data and

Reviewer: ```suggestion
(it holds large derived data). Ignored data is released separately with the manuscript (see the Data and
```

Code Availability section of the paper for the DOI). To reproduce, download the released data
bundle and place it at `data/final_results/` so the default paths resolve. Key files:

| File | Contents |
|---|---|

> | `almanack_metrics.csv` | One row per tool, all Almanack metrics (built from the CCT JSON outputs) |

Reviewer: I highly recommend avoiding the use of CSV for large or complex data handling. Instead, consider using [Parquet](https://parquet.apache.org/), which is well integrated into data science tooling and includes features like strong data typing and columnar access (providing performance benefits). Given that the data have been collected already, perhaps this is work for a future PR or revision.

| `combined_almanack_full_classified_with_joss.csv` | Full joined table: Almanack checks + JOSS scores + CZI mentions + domain labels (the canonical analysis table) |
| `aggregated_data_llm_classified.csv` | `tool_name`, `domain` (LLM-classified) |
| `weights_logistic_refit.csv` | nf-core-calibrated logistic weights per check |
| `nf_core_almanack_metrics.csv` | Almanack metrics for nf-core pipelines |

> A few scripts consume **raw inputs that are not in the data bundle** because they are large or

Reviewer: ```suggestion
Some code consumes **raw inputs that are not in the data bundle** because they are large or
```

regenerable: the per-repo Almanack JSON directory (CCT pipeline output) consumed by

> `build_almanack_metrics_table.py`. Everything needed for the figures is in the released tables.

Reviewer: ```suggestion
`build_almanack_metrics_table.py`. All data needed for the figures is in the released tables.
```

## Pipeline (run order)

```
data_collection/      discover repos, link CZI/PubMed, classify domains
       |
   [ CCT Nextflow pipeline runs each repo -> Almanack + JOSS JSONs ]
       |
data_processing/      build feature table, add literature signals, merge, weight, verify
       |
modeling/             gradient-boosting + SHAP (adoption predictors), domain analysis
       |
visualization/        manuscript figures

> (review-response analyses live separately in ../review_response/)

Reviewer: Consider using [mermaid](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams) to help generate markdown-compatible diagrams which render to visuals in GitHub.

```

## Which script makes which figure / result

| Manuscript item | Script |
|---|---|
| Figure 1 (conceptual framework), Figure 2 (cohort funnel) | `visualization/generate_conceptual_figures.py` |
| Figure 3 (check pass rates), Figure 4 (domain violin) | `visualization/generate_results_figures.py` |
| Figure 5 (SHAP adoption predictors, stars) | `modeling/stars_shap_model.py` |
| Figure 5 mentions panel / robustness | `modeling/mentions_shap_model.py` |
| Figure 6 (longitudinal evolution, grading) | `visualization/benchmark_analysis.py` |
| Section 4.2 domain ANOVA + residuals | `modeling/stars_domain_analysis.py`, `data_processing/apply_logistic_weights_and_domain_anova.py`, `modeling/weighted_scores_statistical_analysis.py` |
| Score distributions (supplementary) | `visualization/plot_almanack_score_distributions.py` |

> | Review-response analyses (supplementary) | `../review_response/` (separate top-level folder) |

Reviewer: Consider avoiding the figure numbering, which is contingent on the manuscript document and has risk of changing (possibly incurring duplicate changes here each time something changes in the manuscript). Instead, we could use exact paths as references within the manuscript if needed.

## Reproducing the headline results

```bash
# Figure 5: sustainability-only adoption predictors (license, citability, modern branch top-ranked)

> python3 analysis/modeling/stars_shap_model.py --sustainability_only

Reviewer: If you opt for `uv`, each of these calls could be converted into `uv run` statements which will automatically invoke the related environment.

# Figure 5 full model (social signals included)
python3 analysis/modeling/stars_shap_model.py

# Domain analysis (Section 4.2)
python3 analysis/modeling/stars_domain_analysis.py

# Figures 3 and 4
python3 analysis/visualization/generate_results_figures.py
```

Outputs land in `data/final_results/` (tables) and `docs/manuscript_drafts/figures/` (figures).

## Review-response analyses

> Analyses added in response to peer review live in the separate top-level `review_response/`

Reviewer: Consider relying on the same analyses directory and code, leveraging git for versioning of the files (we can always reference the history if need-be, respecting that perspectives change over time). This would likely mean less duplication and possibility for confusion for authors or audience (readers would likely be interested in what was published rather than draft content).

Author: I agree! This was initially a folder I created in response to a peer reviewing tool, but decided we didn't need to mention it at all.

folder, kept apart from the main manuscript pipeline. See
[../review_response/README.md](../review_response/README.md).

## Notes

- **Run from the repository root** so relative default paths (`data/final_results/...`) resolve.

> - **API keys** are read from environment variables, never hardcoded: `ANTHROPIC_API_KEY`,

Reviewer: Consider using a dotenv file (e.g. `.env`) for this and using accordingly with [this lightweight wrapper](https://github.com/theskumar/python-dotenv). The way this usually works in practice is there is a `.env_template and a .env file, with the .env file being git-ignored and the .env_template having demo entries but the needed format. This helps to increase reproducibility and expectation-setting for the project in a declarative manner.

  `GITHUB_TOKEN`, and `GEMINI_API_KEY` / `OPENAI_API_KEY` for the alternate classifiers.
- Personal/operational scripts (S3 reparse, shell helpers) and superseded data-assembly steps
  (`recompute_logistic_weights`, `combine_external_datasets`) live in `analysis/archive/` and are
  not needed to reproduce the paper; the tables they would produce are in the released bundle.
- Domains were classified with `data_collection/classify_domains_claude.py` (primary)

>   and `data_collection/classify_batch_bedrock.py` ("Other" bucket). Alternate

Reviewer: Consider naming these files more distinctly to their purpose - perhaps "classify_domains_primary" and "classify_domains_other".

# Analysis code: "Mapping the Garden"

> Reproducible analysis for the manuscript *Mapping the Garden: Software Sustainability as a Correlate of Scientific Impact Across Computational Biology Ecosystems*.

Reviewer: Consider removing the title here to make this work more flexible to changes that might be more strongly correlated with the manuscript draft.

This analysis uses the per-repository measurements produced by the Cancer Complexity Toolkit (CCT) Nextflow pipeline (in the repository root) and produce the statistics and figures in the paper.

## Setup

Dependencies are managed with [uv](https://docs.astral.sh/uv/) via `analysis/pyproject.toml`, and `analysis/uv.lock` pins the exact versions last used for the manuscript.

```bash
# from the repository root
uv sync --project analysis                          # core deps (figures/modeling)
uv sync --project analysis --extra data-collection  # add repo discovery / classification deps
```

Run any code through uv so it uses the locked environment; work **from the repository root** so relative default paths resolve:

```bash
uv run --project analysis python analysis/modeling/stars_shap_model.py --help
```

All code uses argparse with sensible defaults.
Pass `--help` to any script to see its options.

## Data

The code reads from and writes to `data/final_results/`.
That directory is intentionally **ignored** (it holds large derived data).

> Ignored data is released separately with the manuscript (see the Data and Code Availability section of the paper for the DOI).

Reviewer: Consider adding a reference to the dataset within the CITATION.cff to help indicate where folks may access that data. We might also want to provide this information within the readme here too when ready.
